### PR TITLE
Make component removal best-effort and preserve single-error behavior

### LINKED ISSUE #4794: Make the removal of components consistent
Currently, the behavior of `rustup component remove` depends on the order of the provided components.

If an invalid component is listed first, the command fails immediately, and no components are removed.
However, if valid components are listed before an invalid one, those valid components are removed before the command returns an error.

This behavior is thus inconsistent, as it is dependent on the argument ordering.
From a usability perspective, I would argue that it is more intuitive for the command to attempt to remove all valid components and report any invalid ones afterward, regardless of their position.

That said, I am interested in hearing others' thoughts on this.

---
Examples of said inconsistency:

```
$ rustup component remove asd rustc-dev llvm-tools
error: toolchain 'stable-x86_64-unknown-linux-gnu' does not contain component 'asd' for target 'x86_64-unknown-linux-gnu'
```

```
$ rustup component remove rustc-dev llvm-tools asd
info: removing component rustc-dev
info: removing component llvm-tools
error: toolchain 'stable-x86_64-unknown-linux-gnu' does not contain component 'asd' for target 'x86_64-unknown-linux-gnu'
```

### PR DESCRIPTION
**Pull request by Inconnu08**
Previously, `rustup component remove` parsed and removed components in a single pass,
which meant the command stopped on the first error after potentially removing earlier
valid components. As a result, behavior depended on argument order.

This change makes component removal best-effort by:
- first collecting all parseable requested components
- then attempting to remove each of them
- deferring error reporting until the end

This makes the behavior consistent regardless of whether an invalid component appears
before or after valid ones as discussed here: #4794 .

Also adds a regression test covering both orderings.

Closes #4794

### CODE REVIEW

File: src/cli/rustup_mode.rs [hunk 1]
```diff
@@ -1481,12 +1481,33 @@ async fn component_remove(
     let distributable = DistributableToolchain::from_partial(toolchain, cfg).await?;
     let target = get_target(target, &distributable);
 
+    let mut parsed_components = Vec::new();
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: I think it'd not be user friendly enough to only show the first error? I'd expect to get all wrong names at once and only correct once. You can even add/generalize the existing error type for this.

**Review by rami3l — COMMENTED** [review]
Many thanks for working on this :)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/9f7bd9c2780bfed5d08c6d53ce8e196aa7a3a2e5..9bcc18098b766b572d7526890a23e5ef14a95a36)*

File: src/cli/rustup_mode.rs (hunk 1 above)
Comment (Inconnu08) [reply src/cli/rustup_mode.rs]: I wanted to keep the change minimal, but I agree 💯 it would be better to report all invalid component names at once. I’ll update the implementation to collect all failures and use a generalized error type for them.

File: src/cli/rustup_mode.rs [hunk 2]
```diff
@@ -1482,11 +1482,12 @@ async fn component_remove(
     let target = get_target(target, &distributable);
 
     let mut parsed_components = Vec::new();
+    let mut unknown_components = Vec::new();
     let mut first_error = None;
 
     for component in &components {
         match Component::try_new(component, &distributable, target.as_ref()) {
-            Ok(new_component) => parsed_components.push(new_component),
+            Ok(new_component) => parsed_components.push((new_component, component.clone())),
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: Since `component` borrows from `components`, I don't think this `.clone()` is needed; instead you could delay the clone till the moment of generating the error `RustupError::UnknownComponents`.

File: src/cli/rustup_mode.rs [hunk 3]
```diff
@@ -1495,12 +1496,27 @@ async fn component_remove(
...
+    for (component, original_name) in parsed_components {
+        if let Err(err) = distributable.remove_component(component).await {
+            if let Some(RustupError::UnknownComponent { suggestion, .. }) =
+                err.downcast_ref::<RustupError>()
+            {
+                unknown_components.push((original_name, suggestion.clone()));
+                continue;
+            }
+
+            if first_error.is_none() {
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: Nit: This can be more clearly written as `first_error.get_or_insert(err)`.

File: tests/suite/cli_v2.rs [hunk 4]
```diff
@@ -2180,7 +2180,7 @@ async fn remove_component_suggest_best_match() {
         .expect(["rustup", "component", "remove", "rsl"])
         .await
         .with_stderr(snapbox::str![[r#"
-error: toolchain 'nightly-[HOST_TRIPLE]' does not contain component 'rsl' for target '[HOST_TRIPLE]'; did you mean 'rls'?
+error: toolchain 'nightly-[HOST_TRIPLE]' does not contain component 'rsl'; did you mean 'rls'?
```
Comment (rami3l) [inline tests/suite/cli_v2.rs]: Question: This regression seems to be avoidable?

File: src/cli/rustup_mode.rs (hunk 2 above)
Comment (Inconnu08) [reply src/cli/rustup_mode.rs]: Ah, yes ...updated to avoid cloning eagerly and instead clone only when constructing the grouped error

File: src/cli/rustup_mode.rs (hunk 3 above)
Comment (Inconnu08) [reply src/cli/rustup_mode.rs]: Wow, that's cleaner, thanks 👍

File: tests/suite/cli_v2.rs (hunk 4 above)
Comment (Inconnu08) [reply tests/suite/cli_v2.rs]: Yes ...I’ve now updated the implementation to preserve the existing single-component UnknownComponent behavior and only use grouped reporting when multiple invalid components are present.

File: src/cli/rustup_mode.rs [hunk 5]
```diff
@@ -1495,12 +1496,27 @@ async fn component_remove(
         }
     }
 
-    for component in parsed_components {
-        if let Err(err) = distributable.remove_component(component).await
-            && first_error.is_none()
-        {
-            first_error = Some(err);
+    for (component, original_name) in parsed_components {
+        if let Err(err) = distributable.remove_component(component).await {
```
Comment (djc) [inline src/cli/rustup_mode.rs]: Nit: rewrite this condition to avoid rightward drift? Like

```rust
let Err(err) = distributable.remove_component(component).await {
     continue;
};
```

File: src/errors.rs [hunk 6]
```diff
@@ -242,3 +247,27 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest, toolchain: &
 
     String::from_utf8(buf).unwrap()
 }
+
+fn unknown_components_msg(desc: &ToolchainDesc, components: &[(String, Option<String>)]) -> String {
+    let mut buf = vec![];
```
Comment (djc) [inline src/errors.rs]: I don't think it really makes sense to initialize this to a `Vec` and then convert it to `String` at the end? Suggest to make it `String` immediately and use `.write_fmt(format_args!(..))` to write into it.

**Review by djc — COMMENTED** [review]
These commits are really not independent -- would prefer to have them all squashed.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/9f7bd9c2780bfed5d08c6d53ce8e196aa7a3a2e5..6c5bb6e06ff05fd74e0379e13fafe992c5df1059)*

File: src/cli/rustup_mode.rs (hunk 5 above)
Comment (Inconnu08) [reply src/cli/rustup_mode.rs]: Thanks, this reads much cleaner!

File: src/errors.rs (hunk 6 above)
Comment (Inconnu08) [reply src/errors.rs]: True that ...switched to building the message directly as a String 👍

**Review by djc — APPROVED** [review]

**Comment by rami3l** [comment]
Hmmm my apologies, but I think I might need some extra time for a final look at this before merging 🙏

**Review by rami3l — COMMENTED** [review]
I'm feel okay about it globally but I have a few concerns:

- I would like to confirm with @FranciscoTGouveia whether the original intention of https://github.com/rust-lang/rustup/issues/4794 is to aggregate certain errors (which is what I believe) rather than just showing the first error but with the bail-out delayed. IMHO the behavior as implemented in this patch will be very confusing to the user if multiple errors do happen. In this case I suggest that we either aggregate `UnknownComponent*` separately and early bail out for the rest, or actually aggregate all errors (which might be more difficult in terms of display).

- The current distinction of `UnknownComponent` and `UnknownComponents` doesn't seem super necessary since it doesn't prevent the API user from creating the latter with only one `Vec` member.

- The tuples in `UnknownComponents` might be better expressed via a separate struct, but that's a minor implementation detail.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/5100b80375fade4ee4c7c0714c1132fec28319d9..eeb7b06f2369f1b48a80da1281ebd39d80f89d07)*

**Comment by Inconnu08** [comment]
Thanks, that makes sense.

My intent with the current implementation was to:

- preserve the existing UnknownComponent behavior for the single-component case
- aggregate unknown component names only when multiple invalid components are present
- keep the rest of the behavior as close as possible to the existing flow

One of the motivations for introducing `UnknownComponents` was also to avoid repeating the full error message multiple times when several invalid components are provided, which felt a bit noisy UX-wise. The grouped format seemed more concise and easier to read in that case.

That said, I agree the mixed behavior could be confusing if multiple different error types occur. I’m happy to adjust the approach once the intended UX for #4794 is confirmed with @FranciscoTGouveia  whether that means aggregating only `UnknownComponent*` and bailing early for others, or reshaping the error handling more uniformly.

Also, point on UnknownComponent vs UnknownComponents: I introduced the second mainly to avoid regressing the existing single-component behavior, but I’m open to simplifying that if we settle on a cleaner model.

And agreed on the tuple, a small struct would likely make that clearer if we keep the grouped variant.

**Comment by FranciscoTGouveia** [comment]
The original intention of 4794 was indeed to install all possible components and report back _every_ invalid one.

For example, running `rustup component add rust-analyzere rustc-dev llvm-toolse`, should install `rustc-dev` successfully, while returning two separate errors, one for each misspelled component.

On another note, I briefly tested the patch and believe it would be worth adding a test case covering the example above, as this is the current output:
```
$ my-rustup component add rust-analyzere rustc-dev llvm-toolse
error: toolchain '1.90-x86_64-unknown-linux-gnu' does not contain component 'rust-analyzere' for target 'x86_64-unknown-linux-gnu'; did you mean 'rust-analyzer'?
```

**Comment by rami3l** [comment]
> The original intention of 4794 was indeed to install all possible components and report back _every_ invalid one.

@FranciscoTGouveia Thanks for your reply! Given that, my suggestions would be:

- It'd be easier to just aggregate all `UnknownComponent*` errors in the installation process and bail out immediately like before if other kinds of errors do occur.

- On the distinction of `UnknownComponent` vs `UnknownComponents`, I'd say it's more idiomatic to special-case the display format on a unified `UnknownComponents`, rather than doing two `enum` variants. In other words: if the `Vec` is of length 1 we print like the old `UnknownComponent`, otherwise we use the new printing format as introduced here.

- It'd be better if the helper `struct` could be introduced.

@Inconnu08 What do you think?

**Comment by rami3l** [comment]
(Assigned @FranciscoTGouveia for review upon his request.)

File: src/errors.rs [hunk 7]
```diff
@@ -242,3 +247,27 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest, toolchain: &
...
+fn unknown_components_msg(desc: &ToolchainDesc, components: &[(String, Option<String>)]) -> String {
+    let mut buf = String::new();
+
+    match components {
+        [] => panic!("`unknown_components_msg` should not be called with an empty collection"),
+        [(component, suggestion)] => {
+            let _ = write!(
+                buf,
+                "toolchain '{desc}' does not contain component '{component}'{}",
+                suggest_message(suggestion),
```
Comment (FranciscoTGouveia) [inline src/errors.rs]: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/5100b80375fade4ee4c7c0714c1132fec28319d9..eeb7b06f2369f1b48a80da1281ebd39d80f89d07)*

What if we insert a newline before calling `suggest_message`?
Otherwise, we could easily exceed 80 chars :/

File: tests/suite/cli_rustup.rs [hunk 8]
```diff
@@ -2136,6 +2136,145 @@ async fn add_remove_multiple_components() {
...
+
+    for file in &files {
+        let path = PathBuf::from(format!(
+            "toolchains/nightly-{}/{}",
+            this_host_triple(),
+            file
+        ));
+        assert!(!cx.config.rustupdir.has(path.parent().unwrap()));
+    }
+}
```
Comment (FranciscoTGouveia) [inline tests/suite/cli_rustup.rs:2181]: *[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/5100b80375fade4ee4c7c0714c1132fec28319d9..eeb7b06f2369f1b48a80da1281ebd39d80f89d07)*

There should also be a third case in which a valid component appears between invalid ones, ensuring that errors are correctly reported for both invalid components.

File: src/errors.rs (hunk 7 above)
Comment (Inconnu08) [reply src/errors.rs]: Good point. I will update this.

File: tests/suite/cli_rustup.rs (hunk 8 above)
Comment (Inconnu08) [reply tests/suite/cli_rustup.rs:2181]: That’s a useful edge case to cover. I will add it to the test.

**Comment by Inconnu08** [comment]
Apologies for the late response. Thanks to both of you for clarifying the intended behavior :) 

My understanding now is:

- we should process all valid components: best-effort
- aggregate all `UnknownComponent*` errors and report them together
- still bail out immediately for other kinds of errors

Using a single `UnknownComponents` variant with display logic that special-cases the length (to preserve the current single-component/best-match UX) makes sense to me and simplifies the current approach.

The earlier split between `UnknownComponent` and `UnknownComponents ` was mainly to avoid regressing the single-component formatting/suggestion behavior, but I’m happy to unify it now that the intended direction is clear.

I’ll update the implementation accordingly, and I’ll also:

- add a test covering the example (`rust-analyzere rustc-dev llvm-toolse`)
- extend the existing tests with a case where a valid component appears between invalid ones
- adjust the error formatting to move suggestions onto a new line for readability

Let me know if that aligns before I push the changes.

File: src/errors.rs (hunk 7 above)
Comment (Inconnu08) [reply src/errors.rs]: I’ve updated the formatting to move the suggestion onto a new `help:` line. This keeps the line length in check and improves readability. I’ve also updated the affected snapshots accordingly.

File: src/errors.rs
Comment (rami3l) [inline src/errors.rs]: I think given that this is not used a lot elsewhere, we can just remove it to save some redundancy? (Please correct me if I'm wrong.)

File: src/errors.rs [hunk 9]
```diff
@@ -242,3 +254,47 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest, toolchain: &
...
+
+            if let Some(suggestion) = &component.suggestion {
+                let _ = write!(buf, "\nhelp: did you mean '{suggestion}'?");
+            }
+
+            if component.description.contains("rust-std") {
+                let _ = write!(
+                    buf,
+                    "\nnote: not all platforms have the standard library pre-compiled: https://doc.rust-lang.org/nightly/rustc/platform-support.html{}",
+                    if desc.channel == Channel::Nightly {
```
Comment (rami3l) [inline src/errors.rs]: Nit: Since we are not forced to writing in expression-oriented style any more (we are no longer in an attribute), prefer doing this instead:

```rs
let _ = write!(buf, "\nnote: not all platforms have the standard library pre-compiled: https://doc.rust-lang.org/nightly/rustc/platform-support.html");
if desc.channel == Channel::Nightly {
  let _ = write!(buf, "\nhelp: consider using `cargo build -Z build-std` instead");
}
```

**Review by rami3l — APPROVED** [review]
That's much clearer than before, thanks!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/4797/9b6b629804d6e4e8c6baf83b62ad3a052f4286ee..e626a89062264285545d48eada123d9f0b2217b4)*

File: src/errors.rs
Comment (Inconnu08) [reply src/errors.rs]: since we’ve unified on UnknownComponents, the old UnknownComponent variant is redundant. I’ll remove it and update any remaining call sites if needed

File: src/errors.rs (hunk 7 above)
Comment (rami3l) [reply src/errors.rs]: Looks cool to me :)

**Comment by rami3l** [comment]
Thanks again @Kobzol , your newly-enabled automations have made the tracking of subtle changes across rebases much easier 🙏

**Comment by rami3l** [comment]
@Inconnu08 Great work!

There are some test-related stuff that I'll update for you :)

**Comment by Kobzol** [comment]
> Thanks again @Kobzol , your newly-enabled automations have made the tracking of subtle changes across rebases much easier 🙏 

That was actually implemented by @Urgau :)
